### ISSUE TITLE
[RFC] Nested model pages should avoid double id on path

### ISSUE THREAD

**Issue by frankiesardo** [blitz-js/blitz#2413]
### What do you want and why?

When you call `blitz generate pages task --parent project` it creates the following files:

```
app/pages/projects/[projectId]/tasks/[taskId]/edit.tsx
app/pages/projects/[projectId]/tasks/[taskId].tsx
app/pages/projects/[projectId]/tasks/index.tsx
app/pages/projects/[projectId]/tasks/new.tsx
```

I think it's a mistake to rely on a double id (projectId and taskId) for a resource, unless its unique key is a combination of the two keys. Listing and creating tasks requires a `projectId` so the routes are correct as they are, but showing and editing a task only need a `taskId` and nesting it under `projects` creates unnecessary problems.

For example: if you're somewhere else in your application and all you have is the `optionId`, you cannot navigate to the option page unless you also pull its `projectId` from the database, which could be inconvenient if you're deeply nested somewhere else.

It also creates headaches in terms of securing the endpoints: if I have `projectId` and `taskId` in the path I always need to check that that task belong to that `projectId` and throw otherwise. A task that has been reassigned to another project will cause a previously valid url to fail now.

This double id is so unnecessary in fact, that is not used by the mutations at all and you can safely move the show and edit pages somewhere else and everything will work as it is.

In terms of REST design, generally speaking, if Task didn't have an autogenerated `id` and instead it relied on a `@@unique(name, projectId)` constraint, then it makes sense to have `/projects/{projectId}/tasks/{taskName}` in the path. But since we're always creating a model with a `@id @default(autoincrement())` we can simply refer to the task as `/tasks/{taskId}`.

### Possible implementation(s)

When you call `blitz generate pages task --parent project` it should create the following files:

```
app/pages/tasks/[taskId]/edit.tsx
app/pages/tasks/[taskId].tsx
app/pages/projects/[projectId]/tasks/index.tsx
app/pages/projects/[projectId]/tasks/new.tsx
```

The `show` and `edit` pages should be added at the top level. This is fine and shouldn't cause any name clashing because every model name should be unique.

The generated code should be updated to remove `projectId` from the `Routes.ShowTaskPage` and `Routes.EditTaskPage`

### Additional context

I hoped to find a more officially-sounding discussion around this problem but all I could find is this SO thread. It's still an interesting read https://stackoverflow.com/questions/20951419/what-are-best-practices-for-rest-nested-resources

**Comment by flybayer** [comment]
Thanks @frankiesardo, this does make a lot of sense. I'll post this around to get others input. If others agree, I'm good with making this change.

**Comment by waptik** [comment]
I second this suggestion. 
Also with url, it makes it messy if your ids are uuid &co, you'll get a very long(just my opinion).

**Comment by Skn0tt** [comment]
Hi @frankiesardo! Great RFC, some thoughts on it:

> In terms of REST design,

I don't think REST principles are relevant in this case. These paths are about what *users* see, not about closely mirroring the database schema.
Take a look at GitHub URLs, for example. This issue is available at `/blitz-js/blitz/issues/2413`. It contains `/issues`, but that's actually not necessary to show the page! If you went to `/blitz-js/blitz/pull/2413` instead, it'd have redirected you to the same page.
Now you might argue that, in terms of REST design, it's not neccessary to know wether there's a PR or an issue behind some ID. But, in terms of *user experience*, it is incredibly helpful to be able to somewhat guess what's hiding behind a link.

>  I always need to check that that task belong to that projectId and throw otherwise

Do you really need to do this? The worst thing to happen is that somebody mistypes opens `/projects/1/tasks/10` while they actually wanted to open `/projects/2/tasks/10`. The only downside which I could spot here is that now you're serving something under it's non-canonical link, potentially confusing people.

---

Generally speaking, I think we should design our conventions (like default-generated paths) such that they encourage to good-UX-by default.
A non-beginner that's well-versed in schema design and has good knowledge of their problem domain will always be able to override what's been generated for them. It's a matter of renaming a folder.

**Comment by frankiesardo** [comment]
Hey @Skn0tt thanks for your comments! I've added some further consideration.

> I don't think REST principles are relevant in this case. These paths are about what users see, not about closely mirroring the database schema.
Take a look at GitHub URLs, for example. This issue is available at /blitz-js/blitz/issues/2413. It contains /issues, but that's actually not necessary to show the page! If you went to /blitz-js/blitz/pull/2413 instead, it'd have redirected you to the same page.

Absolutely agree on your point about REST principles: we should try and think about the user and developer experience. Your example is exactly what I was mentioning as the exception: 

>  if Task didn't have an autogenerated id and instead it relied on a `@@unique(name, projectId)` constraint, then it makes sense to have `/projects/{projectId}/tasks/{taskName}` in the path

`/issue/1` doesn't make sense on its own. An `Issue` has a compound key of `@@unique(projectName, issueNumber)` so you need them both on the path. Having `project/blitz/issue/1` delivers a great user experience because I only care about issue numbers for this project, not about all the issues ever created in Github (private or public). The way `blitz generate` works is by using a global id so I think it makes more sense to refer to this model with its global id, since it's already in the path.

 > Do you really need to do this? The worst thing to happen is that somebody mistypes opens /projects/1/tasks/10 while they actually wanted to open /projects/2/tasks/10. The only downside which I could spot here is that now you're serving something under it's non-canonical link, potentially confusing people.

Imagine Task 10 is assigned to Project 1 and then gets moved to Project 2. When you call `GET  /projects/1/tasks/10` you need to check that task 10 belongs to project 2 and throw otherwise. If you don't, bad things might happen. 
First of all: it's not true that Task 10 belongs to Project 1 so you shouldn't allow that url to render, otherwise you're confusing people. Then: what if in this `ShowTasksPage` you include a link to go "up" to the parent `Project` via `Routes.ShowProjectPage({ projectId })`? Following it will take you to the wrong project if you're not careful.

___

To summarise, there are two types of `--parent` relationships: a tight and a loose one. A loose relationship is `Member belongs to a Club`. You can create a Member via `/clubs/1/members/new` (so it's assigned to Club 1), but you can later on assign it to Club 2. Why not accessing it via `/members/1` rather than `/clubs/1/members/1` and avoid the headache then?
A tight relationship is `Issue belongs to a Repo`. An Issue only make sense within that Repo. There's no point in moving this Issue to another Repo. It should be fully nested inside of it.

When we generate a model via `blitz generate model --parent` we end up with a loose relationship, because that's just the way the generator works. I don't think there's an easy way to tell Prisma to create a compound key and autoincrement an issue number within the content of a repo. A tight relationship is a more advanced use case and had to be managed manually.

**Comment by Skn0tt** [comment]
Regarding the `/blitz-js/blitz/issues/2413` example, I tried to emphasise on the `/issues/[ticketNumber]` part. Issues, Pull Requests and Discussions are labeled in one consecutive series. If you have their Ticket Number (e.g. 2413 for this issue), it doesn't matter wether you go to `/issues/2413`, `/pulls/2413` or  `/discussions/2413`. Thus, the prefix (`/issues`, `/pulls`, `/discussions) doesn't really matter and could, aligned with your proposal, be omitted. I chose this issue to illustrate why I think it's a good idea to have this redundant information in the URL.

However, I think I understand your point better now. Let me rephrase it:

At the moment, `blitz generate --parent` outputs paths that misleadingly imply they're about entities that only exist in the context of their parent (so-called weak entities). While there's nothing wrong with that, and it's certainly desirable in a lot of cases, the generated database entity isn't actually *weak*! It has its own independent ID space. And that creates confusion, e.g. when moving an entity out of its context (e.g. to a different club) - which the data model happily allows, while the paths look like that shouldn't be possible.

---

There's two ways we can resolve this:

1. Get our paths in-line with the datamodel. What this RFC suggests.
2. Move the datamodel in-line with our paths. aka actually generate weak entities.

Judging by the name `--parent`, I'd expect weak entities to be generated in my data model. That's the idea behind a `parent` node, right? Their children rely on some context (their parents) to make sense of them.

Judging [by our Docs](https://blitzjs.com/docs/cli-generate#model-fields), you could already run `blitz generate model task project:Project`. What's the reason that you need `--parent` to generate your `/tasks/1` paths, @frankiesardo? What am I missing?

**Comment by frankiesardo** [comment]
Thanks for rephrasing both my point and yours, they're much clearer now.

Yes, I think if there was a way to generate a tightly coupled `--parent` relationship via blitz it would be the ideal solution. This is kind of what I was expecting when I hit enter.

The trouble is, I don't think it's that easy and requires a lot of wiring up, so I wasn't disappointed to see the generated model use autoincrement id: I understood the limitations of the tool. 

Problems happen when you have these two global unique ids in the url, so if the second id could be generated relative to the first one, I would be happy about it.

If we keep the model generation the way it is (which is not wrong, just more generic), I think the UX works best as `list` and `created` pages are nested while `show` and `edit` pages are at the root level. Consider this user journey:

```
/projects
/projects/new
/projects/101
/projects/101/tasks
/projects/101/tasks/new
/tasks/42
/tasks/42/edit
```

This works well as a starting point and it's the least surprising behaviour, in my opinion. You expose the end users to how nested paths work (`list` and `create` pages), so they could refactor the other pages if they decided to do so. We could include a more advanced tutorial that builds on this example by adding:

```
model Task {
 ...
 projectId Int
 slug String
 @@unique(slug, projectId)
}
```

And moving the `show` and `edit` pages under `projects`. So the end result will be:

```
/projects
/projects/new
/projects/101
/projects/101/tasks
/projects/101/tasks/new
/projects/101/tasks/my-awesome-task
/projects/101/tasks/my-awesome-task/edit
```

**Comment by flybayer** [comment]
Looks like there are no objections to this change, so let's do it! Ready for someone to work on.
